### PR TITLE
Make coding contract title click-to-copy.

### PR DESCRIPTION
**Pull request by Crownie88**
CopyableText.tsx
 * Added enum ClickableTag which is added to IProps
   so we can say what tag should be use to wrap the clickable text

CodingContractPopup.tsx
 * Popup for displaying contract info + input fields
   Stolen from HashUpgrade popups

PopupButton.tsx
 * generic button class allowing your own clickhandler to be given via props

PopupCloseButton.tsx
 * changed it so it now inherits from popupbutton

CodingContracts.ts
 * the prompt function now calls on CodingContractPopup to generate the window
 * removed unused imports

### CODE REVIEW

File: src/CodingContracts.ts [hunk 1]
```diff
@@ -130,6 +129,8 @@ export class CodingContract {
 
     /* String representing the contract's type. Must match type in ContractTypes */
     type: string;
+    /* The id used for the popup. We need this to close it later. */
+    popupId: string;
```
Comment (hydroflame) [inline src/CodingContracts.ts:133]: This variable is only used inside `prompt` so should be declared there too.

File: src/CodingContracts.ts [hunk 2]
```diff
@@ -147,6 +148,8 @@ export class CodingContract {
         this.type = type;
         this.data = CodingContractTypes[type].generate();
         this.reward = reward;
+        this.popupId = `coding-contract-prompt-popup-${this.fn}`;
+        // this.popup = new CodingContractPopup({c:this});
```
Comment (hydroflame) [inline src/CodingContracts.ts:152]: Remove comment

File: src/CodingContracts.ts [hunk 3]
```diff
@@ -173,63 +176,30 @@ export class CodingContract {
         return CodingContractTypes[this.type].solver(this.data, solution);
     }
 
+    isCancelled() {
```
Comment (hydroflame) [inline src/CodingContracts.ts:179]: unused, please remove

File: src/ui/CodingContractPopup.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import { PopupCloseButton } from "../ui/React/PopupCloseButton";
+
+type IProps = {
+    c: CodingContract;
+    onCloseClick: () => void;
+    onSolveClick: () => void;
+}
+
+export class CodingContractPopup extends React.Component<IProps>{
+    state: any;
```
Comment (hydroflame) [inline src/ui/CodingContractPopup.tsx:15]: state should not be `any` type

File: src/ui/CodingContractPopup.tsx [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import * as React from "react";
+import { KEY } from "../../utils/helpers/keyCodes";
+
+import { CodingContract, CodingContractType, CodingContractTypes } from "../CodingContracts";
+import { ClickableTag, CopyableText } from "../ui/React/CopyableText";
+import { PopupCloseButton } from "../ui/React/PopupCloseButton";
+
+type IProps = {
+    c: CodingContract;
+    onCloseClick: () => void;
```
Comment (hydroflame) [inline src/ui/CodingContractPopup.tsx:10]: To match naming convention these should be named `onClick` and `onSolve`.

also it's an `attempt` not a `solve`

File: src/ui/CodingContractPopup.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+export class CodingContractPopup extends React.Component<IProps>{
+    state: any;
+    constructor(props: IProps) {
+        super(props);
+        this.state = { value: ''};
+        this.setValue = this.setValue.bind(this);
+    }
+    getValue(){
+        let input = document.getElementById('contractInput')?.getAttribute('value');
```
Comment (hydroflame) [inline src/ui/CodingContractPopup.tsx:22]: See https://github.com/danielyxie/bitburner/blob/master/src/Faction/ui/DonateOption.tsx for example on how to bind input value to state.

File: src/ui/CodingContractPopup.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+            description.push(<span key={i}>{value}<br/></span>);
+        return (
+            <div>
+                <CopyableText value={this.props.c.type} tag={ClickableTag.Tag_h1} />
+                <br/><br/>
+                <p>You are attempting to solve a Coding Contract. You have {this.props.c.getMaxNumTries() - this.props.c.tries} tries remaining, after which the contract will self-destruct.</p>
+                <br/>
+                <p>{description}</p>
+                <br/>
+                <input style={{ width:"50%",marginTop:"8px" }} autoFocus={true} placeholder="Enter Solution here" value={this.state.value}
```
Comment (hydroflame) [inline src/ui/CodingContractPopup.tsx:53]: This input element style should match other text inputs, like exp inputs under the dev menu.

File: src/ui/React/PopupButton.tsx [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+import * as React       from "react";
+import *  as ReactDOM   from "react-dom";
+
+import { KEY } from "../../../utils/helpers/keyCodes";
+import { removeElement } from "../../../utils/uiHelpers/removeElement";
+
+export interface IPopupButtonProps {
+    class?: string;
+    popup: HTMLElement | string;
+    ClickHandler: () => void;
```
Comment (hydroflame) [inline src/ui/React/PopupButton.tsx:17]: `onClick`

File: src/CodingContracts.ts [hunk 9]
```diff
@@ -175,31 +171,29 @@ export class CodingContract {
...
+                popupId: `coding-contract-prompt-popup-${this.fn}`,
+                onClose: () => {
                     resolve(CodingContractResult.Cancelled); 
-                    removePopup(this.popupId);
+                    removePopup(popup.props.popupId);
                 },
-                onSolveClick: () => {
-                    const answer: string = popup.getValue();
+                onAttempt: (val: string) => {
+                    alert(val);
```
Comment (Crownie88) [inline src/CodingContracts.ts:186]: Remove alert you idiot

File: src/CodingContracts.ts [hunk 10]
```diff
@@ -175,31 +171,29 @@ export class CodingContract {
...
+                    const answer: string = popup.state.value;
```
Comment (Crownie88) [inline src/CodingContracts.ts:187]: This can be removed. Use argument of event to check the answer

File: src/CodingContracts.ts [hunk 11]
```diff
@@ -175,31 +171,29 @@ export class CodingContract {
...
                     if (this.isSolution(answer)) {
                         resolve(CodingContractResult.Success);
                     } else {
-                        resolve(CodingContractResult.Failure);
+                        resolve(CodingContractResult.Cancelled); //resolve(CodingContractResult.Failure);
```
Comment (Crownie88) [inline src/CodingContracts.ts:191]: This isnt commited but was reverted

**Comment by hydroflame** [comment]
Copied and merged Thanks!

**Pull request closed without merging**